### ISSUE TITLE
GPU support in Panzer: failing tests

### ISSUE THREAD

**Issue by eric-c-cyr** [trilinos/Trilinos#318]
This issue is to track evolving panzer support on GPUs. With the @etphipp workaround of #227, we can make progress on panzer. After that commit there are a few issues with regard to failing tests on the GPU. Only the STKClassic tests are failing. The set of failing tests/examples is:

PanzerAdaptersSTK_assembly_engine_MPI_4: Timeout...
PanzerAdaptersSTK_response_residual_MPI_2
PanzerAdaptersSTK_CurlLaplacianExample
PanzerAdaptersSTK_CurlLaplacianExample-ConvTest
PanzerAdaptersSTK_MixedPoissonExample
PanzerAdaptersSTK_MixedPoissonExample-ConvTest
PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4
PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4-ConvTest
PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest
PanzerAdaptersSTK_main_driver_energy-ss-loca-eigenvalue

**Comment by eric-c-cyr** [comment]
After looking at these tests a bit more carefully the assembly_engine test needs to be rewritten completely for GPUs. The GPU is initialized, memory allocated and stored, and the GPU is then uninitialized. Then the memory is accessed again after the GPU is reinitialized. Bad things happen.

**Comment by eric-c-cyr** [comment]
The tests
PanzerAdaptersSTK_CurlLaplacianExample
PanzerAdaptersSTK_CurlLaplacianExample-ConvTest
PanzerAdaptersSTK_MixedPoissonExample
PanzerAdaptersSTK_MixedPoissonExample-ConvTest

All fail in the panzer::Sum evaluator. If I set the macro PANZER_USE_FAST_SUM to '1' then they pass.

**Comment by eric-c-cyr** [comment]
The  PanzerAdaptersSTK_response_residual_MPI_2 test was missing tolerances. I've loosened them slightly. I will push this change shortly.

**Comment by mhoemmen** [comment]
> With the @etphipp workaround of #227, we can make progress on panzer. After that commit there are a few issues with regard to failing tests on the GPU.

Please note that Tpetra does not yet pass all its tests with CUDA.  I'm working on a more robust fix.  Thanks though to @etphipp for the work-around :-)

**Comment by eric-c-cyr** [comment]
I've pushed a large change to all the tests to rely on the phalanx unit test driver. This initializes Kokkos only once. This was necessary to run on different platforms. There are still outstanding test failures in panzer that we are working through.

**Comment by eric-c-cyr** [comment]
The tests appear to pass at this point. However, two examples are failing. These PoissonInterfaceExample fails with an Epetra error.

**Comment by eric-c-cyr** [comment]
Regarding the failing examples, I wonder if the epetra scatter is to blame here and the use of the "dx" pointer.

**Comment by mhoemmen** [comment]
Just to note -- #227 has been fixed (not just worked-around) since I last wrote.

**Comment by mhoemmen** [comment]
I'm closing this issue since it appears that everything is fine now.  Please reopen or open a new issue if this is still happening.
